### PR TITLE
Add jsx-sort-default-props rule

### PR DESCRIPTION
**Pull request by b0gok**
Related to https://github.com/yannickcr/eslint-plugin-react/issues/281

- [x] Add support for
```js
var First = createReactClass({
  getDefaultProps: function() {
    return {
      barRequired: "barRequired",
      onBar: "onBar"
    };
  }
});
```
- [x] Add more tests
- [x] Remove useless code

### CODE REVIEW

File: docs/rules/jsx-sort-default-props.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Enforce propTypes declarations alphabetical sorting (react/jsx-sort-default-props)
+
+Some developers prefer to sort propTypes declarations alphabetically to be able to find necessary declaration easier at the later time. Others feel that it adds complexity and becomes burden to maintain.
```
Comment (ljharb) [inline docs/rules/jsx-sort-default-props.md]: let's make sure all `propTypes` and `defaultProps` mentions are surrounded by backticks.

Also, "to be able to find necessary declarations easier at a later time" and "becomes a burden to maintain"

File: docs/rules/jsx-sort-default-props.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+## Rule Details
+
+This rule checks all components and verifies that all defaultProps declarations are sorted alphabetically. A spread attribute resets the verification. The default configuration of the rule is case-sensitive.
```
Comment (ljharb) [inline docs/rules/jsx-sort-default-props.md]: Can you elaborate about the spread attribute?

I would expect that sorts do not cross a spread boundary, but `d, a, ...obj, c, b` would be sorted into `a, d, ...obj, b, c`

**Comment by fsmaia** [comment]
Any advances? I have some spare time, may I help?

**Comment by b0gok** [comment]
@fsmaia sorry, haven't free time for a while. So help is welcome :)

**Comment by ljharb** [comment]
@b0gok the ideal is for you to add @fsmaia to your fork, so they can update this PR directly - if not, then @fsmaia, please post here a link to your commits on your own fork, and I'll add them for you. Please do *not* open up a duplicate PR.

**Comment by b0gok** [comment]
@ljharb ok.
@fsmaia I add you as collaborator in my fork.

File: docs/rules/jsx-sort-default-props.md (hunk 1 above)
Comment (fsmaia) [reply docs/rules/jsx-sort-default-props.md]: Done! Should I fix these typos in `sort-prop-types` and `jsx-sort-props` documentation too? Their introduction are identical.

**Comment by fsmaia** [comment]
Added some SFC examples in the documentation and tests. Had to change the rule source, 'cause unnecessary/commented code was useful right in SFC scenario.

**Comment by fsmaia** [comment]
@ljharb: Are we ready to merge? Any further work or change needed? Thanks

File: docs/rules/jsx-sort-default-props.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+const Component = (props) => (...);
+Component.defaultProps = {
+  z: "z",
+  y: "y",
+  a: "a"
+};
+```
+
+The following patterns are considered okay and do not cause warnings:
```
Comment (ljharb) [inline docs/rules/jsx-sort-default-props.md]: Is the wording of this sentence consistent with the other rules?

**Comment by ljharb** [comment]
Also, please rebase and force push this so as to remove merge commits.

**Comment by b0gok** [comment]
> Can we add some good and bad examples in the docs that use spreads?

@ljharb you meen something like this?
```js
const defaults = {
  foo: "foo"
};
const types = {
  foo: PropTypes.string,
  bar: PropTypes.string
};

function MyStatelessComponent({ foo, bar }) {
  return <div>{foo}{bar}</div>;
}

MyStatelessComponent.propTypes = types;
MyStatelessComponent.defaultProps = {
  ...defaults,
  bar: "bar"
};
```

**Comment by ljharb** [comment]
Yes, just like that.

**Comment by b0gok** [comment]
Added more tests with spreads and examples.

**Comment by EvHaus** [comment]
Just curious, why not update the existing `jsx-sort-props` rule to sort defaultProps instead of create a separate rule? Is there a situations where you'd want proptypes sorted one way, but defaults sorted a different way?

**Pull request merged**